commit 94389346492346be1fad619552e3ea0ce26b4afc
Author: Dale Rahn <drahn@dalerahn.com>
Date:   Sun Aug 11 22:14:20 2013 -0400

    Important differences between FreeBSD clang and patched bitrig-llvm

diff --git a/contrib/llvm/lib/Support/Unix/Host.inc b/contrib/llvm/lib/Support/Unix/Host.inc
index 7f79db0..0191335 100644
--- a/contrib/llvm/lib/Support/Unix/Host.inc
+++ b/contrib/llvm/lib/Support/Unix/Host.inc
@@ -26,11 +26,11 @@
 
 using namespace llvm;
 
-#ifdef __FreeBSD__
+#if defined(__FreeBSD__) || defined(__Bitrig__)
 std::string sys::getDefaultTargetTriple() {
   return LLVM_DEFAULT_TARGET_TRIPLE;
 }
-#else // __FreeBSD__
+#else // __FreeBSD__ ||  __Bitrig__
 static std::string getOSVersion() {
   struct utsname info;
 
diff --git a/contrib/llvm/tools/clang/include/clang/Basic/DiagnosticGroups.td b/contrib/llvm/tools/clang/include/clang/Basic/DiagnosticGroups.td
index b0d0216..156e9ed 100644
--- a/contrib/llvm/tools/clang/include/clang/Basic/DiagnosticGroups.td
+++ b/contrib/llvm/tools/clang/include/clang/Basic/DiagnosticGroups.td
@@ -63,6 +63,7 @@ def : DiagGroup<"ctor-dtor-privacy">;
 def GNUDesignator : DiagGroup<"gnu-designator">;
 def GNUStringLiteralOperatorTemplate :
   DiagGroup<"gnu-string-literal-operator-template">;
+def : DiagGroup<"bounded">;
 
 def DeleteNonVirtualDtor : DiagGroup<"delete-non-virtual-dtor">;
 def AbstractFinalClass : DiagGroup<"abstract-final-class">;
@@ -498,6 +499,7 @@ def Most : DiagGroup<"most", [
     StringPlusInt,
     Trigraphs,
     Uninitialized,
+    UninitializedSometimes,
     UnknownPragmas,
     Unused,
     VolatileRegisterVar,
diff --git a/contrib/llvm/tools/clang/lib/Basic/Targets.cpp b/contrib/llvm/tools/clang/lib/Basic/Targets.cpp
index bccd0d7..8a2211f 100644
--- a/contrib/llvm/tools/clang/lib/Basic/Targets.cpp
+++ b/contrib/llvm/tools/clang/lib/Basic/Targets.cpp
@@ -430,6 +430,7 @@ protected:
     // Bitrig defines; list based off of gcc output
 
     Builder.defineMacro("__Bitrig__");
+    Builder.defineMacro("__OpenBSD__");
     DefineStd(Builder, "unix", Opts);
     Builder.defineMacro("__ELF__");
     if (Opts.POSIXThreads)
@@ -438,7 +439,7 @@ protected:
 public:
   BitrigTargetInfo(const llvm::Triple &Triple) : OSTargetInfo<Target>(Triple) {
     this->UserLabelPrefix = "";
-    this->TLSSupported = false;
+    this->TLSSupported = true;
     this->MCountName = "__mcount";
   }
 };
